File: main/web/src/collector.py
import aiohttp
import random
import sys
import asyncio
import json


apic_services = {
  'proxyscrape': 'https://api.proxyscrape.com/v4/free-proxy-list/get?request=display_proxies&proxy_format=protocolipport&format=json',
  'proxifly': 'https://cdn.jsdelivr.net/gh/proxifly/free-proxy-list@main/proxies/all/data.json',
  'jetkai-proxy-list': 'https://raw.githubusercontent.com/jetkai/proxy-list/main/online-proxies/json/proxies-advanced.json',
  'monosans-proxy-list': 'https://raw.githubusercontent.com/monosans/proxy-list/main/proxies_pretty.json',
  'vakhov-proxy-list': 'https://raw.githubusercontent.com/vakhov/fresh-proxy-list/refs/heads/master/proxylist.json',
  'geonode': [
    'https://proxylist.geonode.com/api/proxy-list?limit=500&page=1&sort_by=lastChecked&sort_type=desc',
    'https://proxylist.geonode.com/api/proxy-list?limit=500&page=2&sort_by=lastChecked&sort_type=desc',
    'https://proxylist.geonode.com/api/proxy-list?limit=500&page=3&sort_by=lastChecked&sort_type=desc'
  ]
}

user_agents = [
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:120.0) Gecko/20100101 Firefox/120.0',
  'Mozilla/5.0 (Windows NT 11.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36 Edg/121.0.0.0',
  'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
  'Mozilla/5.0 (X11; Linux x86_64; rv:120.0) Gecko/20100101 Firefox/120.0',
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 13_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.6 Safari/605.1.15',
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 14_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36',
  'Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36',
  'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:102.0) Gecko/20100101 Firefox/102.0'
]


async def request(name: str, url: str):
  try:
    async with aiohttp.ClientSession(headers={'User-Agent': random.choice(user_agents)}) as session:
      async with session.get(url) as response:
        if response.status == 200:
          text = await response.text()
          data = json.loads(text)
          return { 'success': True, 'name': name, 'data': data }
        else:
          return { 'success': False, 'name': name, 'error': 'Некорректный код ответа' }
  except Exception as e:
    return { 'success': False, 'name': name, 'error': str(e) }
  

async def createTaskFlow(urls: dict[str, str]):
  tasks = []

  for name, url in urls.items():
    if name == 'geonode':
      for u in url:
        tasks.append(request(name, u))
    else:
      tasks.append(request(name, url))

  return tasks
  

async def APIc(protocol: str, country: str):
  tasks = await createTaskFlow(apic_services)

  results = await asyncio.gather(*tasks)

  proxies = []

  for result in results:
    if result['success']:
      data = result['data']

      if result['name'] == 'proxyscrape':
        for proxy in data['proxies']:
          if country == 'any' or proxy['ip_data']['countryCode'] == country.upper():
            if protocol == 'any' or str(proxy['protocol']).lower() == protocol:
              proxies.append(proxy['proxy'])
      elif result['name'] == 'proxifly':
        for proxy in data:
          if country == 'any' or proxy['geolocation']['country'] == country.upper():
            if protocol == 'any' or str(proxy['protocol']).lower() == protocol:
              proxies.append(proxy['proxy'])
      elif result['name'] == 'jetkai-proxy-list':
        for proxy in data:
          if country == 'any' or proxy['location']['isocode'] == country.upper():
            if protocol == 'any' or str(proxy['protocols'][0]['type']).lower() == protocol:
              proxies.append(f"{proxy['protocols'][0]['type']}://{proxy['ip']}:{proxy['port']}")
      elif result['name'] == 'monosans-proxy-list':
        for proxy in data:
          if proxy['username'] == None and proxy['password'] == None:
            if country == 'any' or proxy['geolocation']['country']['iso_code'] == country.upper():
              if protocol == 'any' or str(proxy['protocol']).lower() == protocol:
                proxies.append(f"{proxy['protocol']}://{proxy['host']}:{proxy['port']}")
      elif result['name'] == 'vakhov-proxy-list':
        for proxy in data:
          if country == 'any' or proxy['country_code'] == country.upper():
            if protocol == 'any':
              protocols = ['socks5', 'socks4', 'http']
              for p in protocols:
                if int(proxy[p]) != 0:
                  proxies.append(f"{p}://{proxy['ip']}:{proxy['port']}")
            else:
              if int(proxy[protocol]) != 0:
                proxies.append(f"{protocol}://{proxy['ip']}:{proxy['port']}")
      elif result['name'] == 'geonode':
        for proxy in data['data']:
          if country == 'any' or proxy['country'] == country.upper():
            if protocol == 'any':
              protocols = [p.lower() for p in proxy['protocols']]

              for p in protocols:
                proxies.append(f"{p}://{proxy['ip']}:{proxy['port']}")
            else:
              protocols = [p.lower() for p in proxy['protocols']]

              for p in protocols:
                if p == protocol:
                  proxies.append(f"{p}://{proxy['ip']}:{proxy['port']}")

  return proxies


async def file(options: dict[str, any]):
  try:
    if options['mode'] == 'write':
      with open(options['path'], 'w') as f:
        if isinstance(options['data'], dict):
          f.write(json.dumps(options['data']))
        else:
          f.write(str(options['data']))

        f.close()
    elif options['mode'] == 'read':
      result = ''

      with open(options['path'], 'r') as f:
        result += f.read()
        f.close()

      return result
    
  except Exception:
    return None


async def collect(options: dict[str, str]):
  try:
    latest_options = await file({
      'mode': 'read',
      'path': './collector.options.txt'
    })

    if latest_options:
      latest_result = await file({
          'mode': 'read',
          'path': './collector.result.txt'
        })

      if options == json.loads(latest_options) and latest_result != None:
        return latest_result

    await file({
      'mode': 'write',
      'path': './collector.options.txt',
      'data': options
    })

    proxy_list = []

    if options['algorithm'] == 'apic':
      proxy_list = await APIc(options['protocol'], options['country'])

    if options['count'] != 'max':
      proxy_list = proxy_list[:int(options['count'])]

    proxy_list = [str(p) for p in dict.fromkeys(proxy_list)]

    string = '\n'.join(proxy_list)

    await file({
      'mode': 'write',
      'path': './collector.result.txt',
      'data': string
    })

    return string
    
  except Exception as err:
    return err


async def main():
  args = sys.argv

  algorithm = args[1]
  protocol = args[2]
  country = args[3]
  count = args[4]

  result = await collect({ 
    'algorithm': algorithm,
    'protocol': protocol,
    'country': country,
    'count': count
  })

  print(result)

    
if __name__ == '__main__':
  asyncio.run(main())

File: main/src/collector.ts
import { randomInt } from 'crypto';

interface Options { 
  algorithm: string; 
  protocol: string; 
  country: string; 
  count: string;
};

const apicServices = {
  'json:proxyscrape': 'https://api.proxyscrape.com/v4/free-proxy-list/get?request=display_proxies&proxy_format=protocolipport&format=json',
  'json:proxifly': 'https://cdn.jsdelivr.net/gh/proxifly/free-proxy-list@main/proxies/all/data.json',
  'json:jetkai-proxy-list': 'https://raw.githubusercontent.com/jetkai/proxy-list/main/online-proxies/json/proxies-advanced.json',
  'json:monosans-proxy-list': 'https://raw.githubusercontent.com/monosans/proxy-list/main/proxies_pretty.json',
  'json:vakhov-proxy-list': 'https://raw.githubusercontent.com/vakhov/fresh-proxy-list/refs/heads/master/proxylist.json',
  'json:geonode': [
    'https://proxylist.geonode.com/api/proxy-list?limit=500&page=1&sort_by=lastChecked&sort_type=desc',
    'https://proxylist.geonode.com/api/proxy-list?limit=500&page=2&sort_by=lastChecked&sort_type=desc',
    'https://proxylist.geonode.com/api/proxy-list?limit=500&page=3&sort_by=lastChecked&sort_type=desc',
  ]
};

const upcServices = {
  ...apicServices,
  'text:gfpcom-proxy-list': [
    'https://raw.githubusercontent.com/wiki/gfpcom/free-proxy-list/lists/http.txt',
    'https://raw.githubusercontent.com/wiki/gfpcom/free-proxy-list/lists/socks4.txt',
    'https://raw.githubusercontent.com/wiki/gfpcom/free-proxy-list/lists/socks5.txt',
    'https://raw.githubusercontent.com/wiki/gfpcom/free-proxy-list/lists/ss.txt'
  ]
};

const userAgents = [
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:120.0) Gecko/20100101 Firefox/120.0',
  'Mozilla/5.0 (Windows NT 11.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36 Edg/121.0.0.0',
  'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
  'Mozilla/5.0 (X11; Linux x86_64; rv:120.0) Gecko/20100101 Firefox/120.0',
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 13_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.6 Safari/605.1.15',
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 14_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36',
  'Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36',
  'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:102.0) Gecko/20100101 Firefox/102.0',
];

function randomUserAgent() {
  return String(userAgents[randomInt(0, userAgents.length)]);
}

async function request(name: string, url: string, type: string) {
  try {
    const res = await fetch(url, { headers: { 'User-Agent': randomUserAgent() } });
    if (res.ok) {
      const data = type === 'json' ? await res.json() : await res.text();
      return { success: true, name, data };
    } else {
      return { success: false, name, error: `HTTP ${res.status}` };
    }
  } catch (error) {
    return { success: false, name, error: String(error) };
  }
}

function createTaskFlow(urls: any) {
  const tasks = [];

  for (const [name, url] of Object.entries(urls)) {
    if (Array.isArray(url)) {
      for (const u of url) tasks.push(request(String(name.split(':')[1]), u, String(name.split(':')[0])));
    } else {
      tasks.push(request(String(name.split(':')[1]), url as string, String(name.split(':')[0])));
    }
  }

  return tasks;
}

async function UPc(protocol: string) {
  let proxies = [];

  const apicList = await APIc(protocol, 'any');

  if (apicList.length > 0) proxies.push(...apicList);

  const tasks = createTaskFlow(upcServices);
  const results = await Promise.all(tasks);

  for (const r of results) {
    if (!r.success) continue;
    const data = String(r.data).trim().split('\n');
    const name = r.name;

    if (name === 'gfpcom-proxy-list') {
      for (const proxy of data ?? []) {
        if (protocol === 'any' || String(proxy.split('://')[0]).toLowerCase() === protocol) proxies.push(proxy);
      }
    }
  } 

  return proxies;
}

async function APIc(protocol: string, country: string) {
  const tasks = createTaskFlow(apicServices);
  const results = await Promise.all(tasks);

  const proxies = [];

  for (const r of results) {
    if (!r.success) continue;
    const data = r.data;
    const name = r.name;

    if (name === 'proxyscrape') {
      for (const proxy of data.proxies ?? []) {
        if (country === 'any' || proxy.ip_data?.countryCode === country.toUpperCase()) {
          if (protocol === 'any' || String(proxy.protocol).toLowerCase() === protocol) proxies.push(proxy.proxy);
        }
      }
    } else if (name === 'proxifly') {
      for (const proxy of data ?? []) {
        if (country === 'any' || proxy.geolocation?.country === country.toUpperCase()) {
          if (protocol === 'any' || String(proxy.protocol).toLowerCase() === protocol) proxies.push(proxy.proxy);
        }
      }
    } else if (name === 'jetkai-proxy-list') {
      for (const proxy of data ?? []) {
        if (country === 'any' || proxy.location?.isocode === country.toUpperCase()) {
          const type = proxy.protocols?.[0]?.type;
          if (protocol === 'any' || String(type).toLowerCase() === protocol) {
            proxies.push(`${type}://${proxy.ip}:${proxy.port}`);
          }
        }
      }
    } else if (name === 'monosans-proxy-list') {
      for (const proxy of data ?? []) {
        if (proxy.username == null && proxy.password == null) {
          if (country === 'any' || proxy.geolocation?.country?.iso_code === country.toUpperCase()) {
            if (protocol === 'any' || String(proxy.protocol).toLowerCase() === protocol) {
              proxies.push(`${proxy.protocol}://${proxy.host}:${proxy.port}`);
            }
          }
        }
      }
    } else if (name === 'vakhov-proxy-list') {
      for (const proxy of data ?? []) {
        if (country === 'any' || proxy.country_code === country.toUpperCase()) {
          if (protocol === 'any') {
            for (const p of ['socks5', 'socks4', 'http']) {
              if (Number(proxy[p]) !== 0) proxies.push(`${p}://${proxy.ip}:${proxy.port}`);
            }
          } else {
            if (Number(proxy[protocol]) !== 0) proxies.push(`${protocol}://${proxy.ip}:${proxy.port}`);
          }
        }
      }
    } else if (name === 'geonode') {
      for (const proxy of data.data ?? []) {
        if (country === 'any' || proxy.country === country.toUpperCase()) {
          const protocols = (proxy.protocols ?? []).map((p: string) => p.toLowerCase());
          if (protocol === 'any') {
            for (const p of protocols) proxies.push(`${p}://${proxy.ip}:${proxy.port}`);
          } else {
            for (const p of protocols) if (p === protocol) proxies.push(`${p}://${proxy.ip}:${proxy.port}`);
          }
        }
      }
    }
  }

  return proxies;
}

export async function collect(options: Options) {
  try {
    let proxy_list: any[] = [];

    if (options.algorithm === 'apic') {
      proxy_list = await APIc(options.protocol, options.country);
    } else if (options.algorithm === 'upc') {
      proxy_list = await UPc(options.protocol);
    }

    if (options.count !== 'max') {
      proxy_list = proxy_list.slice(0, Number(options.count));
    }

    proxy_list = Array.from(new Set(proxy_list.map(String)));

    const out = proxy_list.join('\n');

    return out;
  } catch (err: any) {
    return err;
  }
}